### ISSUE TITLE
Saving page in admin removes page from hierarchy

### ISSUE THREAD

**Issue by maxinacube** [magento/magento2#9971]
### Preconditions
1. Magento EE 2.1.7
2. PHP 7.0.16
3. MySQL 5.5.45-37.4
4. Active migration from 1.13.0.1 to 2.1.3 and upgraded since

### Steps to reproduce
1. Create a Hierarchy (Ex. Helpful Links)
2. Create a Page (Ex. Contact)
3. Check off desired Hierarchy (Ex. Make Contact a child of Helpful Links)
4. Save Page
5. Edit Contact page and save again

### Expected result
1. Page should save as normal and retain Hierarchy

### Actual result
1. Page saves but is removed from Hierarchy

It _seems_ this happens only if I haven't scrolled open the Hierarchy section on the page I'm editing - when I edit the page with Hierarchy section scrolled open the page doesn't ever seem to be removed from the Hierarchy.

**Comment by veloraven** [comment]
@maxinacube thank you for your report.
Please report EE issues via the Support portal of your account or Partner portal if you are a partner reporting on behalf of a merchant. 
Github is intended for Community edition reports given no account management for CE users. This will allow for proper tracking of issues at the account level.

**Comment by iglaze-kadro** [comment]
I'm having the same problem (Magento version EE 2.1.5, PHP 7), and I'm not sure how to fix it, but here is what I have discovered regarding the problem:
Has to do with the absence of a hidden input field with the id "#nodes_data" on an edit CMS page page (in the admin), which only shows up if the hierarchy section is expanded. In another site where this works (uses an earlier version of Magento2), jQuery("#nodes_data").size() always gives 1. On the site I'm having the issue with (the one I'm currently working on), jQuery("#nodes_data").size() gives 1 only if the Hierarchy menu on the admin's edit CMS page page is expanded, otherwise it gives 0.
The JSON information obtained from "#nodes_data" is important so that the data obtained from vendor/magento/module-cms/Controller/Adminhtml/Page/Save.php in this line of code: execute()
{ $data = $this->getRequest()->getPostValue(); ...... }
includes the JSON value of "#nodes_data".

**Comment by Skullsneeze** [comment]
Glad to say that thanks to @iglaze-kadro 's research I've managed to find a temporary fix until this issues is resolved. What fixed this for me was the following:

1. Override the Magento_VersionsCms module (if you didn't yet)
2. Create the following file within your module override `view/adminhtml/ui_component/cms_page_form.xml`
3. Paste the following code in the file you just created

```
<?xml version="1.0" encoding="UTF-8"?>
<form xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_Ui:etc/ui_configuration.xsd">
    <fieldset name="hierarchy">
        <argument name="data" xsi:type="array">
            <item name="config" xsi:type="array">
                <item name="collapsible" xsi:type="boolean">false</item>
            </item>
        </argument>
    </fieldset>
</form>
```

This disables the collapsible, which is what causes this issue.

For the dev's; The exact cause for this issue is the fact that the content for the collapsible fieldset is only loaded once the fieldset is opened. This works in most cases, but when inline javascript is used this will not be initialized untill the fieldset is opened.

Let me know if this resolves the issue for you @iglaze-kadro and @maxinacube

**Comment by magento-engcom-team** [comment]
@maxinacube thank you for your report. Please follow these guidelines for proper tracking of your issue. You can report Commerce-related issues in one of two ways:
You can use the Support portal associated with your account
or
If you are a Partner reporting on behalf of a merchant, use the Partner portal.
 
GitHub is intended for Magento Open Source users to report on issues related to Open Source only. There are no account management services associated with GitHub.

**Comment by larsroettig** [comment]
@maxinacubeI have analysed this ticket open pr today in the partner repository if you no access write my a Mail to **magentoconsulting@techdivision.com** i send you a patch file

**Comment by evs-xsarus** [comment]
Seems to be fixed in Magento EE 2.3.1 but another problem can arise when the labels are null. This patche fixes it (and also shows the ID in the label):

```diff
--- vendor/magento/module-versions-cms/view/adminhtml/templates/page/tab/hierarchy.phtml
+++ vendor/magento/module-versions-cms/view/adminhtml/templates/page/tab/hierarchy.phtml
@@ -76,9 +76,10 @@
                     var dd = (this.nodes[i].parent_node_id && this.nodes[i].current_page) ? true : false;
                     var cls = this.nodes[i].current_page ? 'cms-current' : '';
                     cls += this.nodes[i].page_id ? ' cms_page' : ' cms_node';
+                    var label = (typeof(this.nodes[i].label) === 'string' ? this.nodes[i].label : '') + ' (ID: ' + this.nodes[i].node_id + ')';
                     var node = new Ext.tree.TreeNode({
                         id: this.nodes[i].node_id,
-                        text: this.nodes[i].label.escapeHTML().replace('\'', '&#039;').replace('"', '&quot;'),
+                        text: label.trim().escapeHTML().replace('\'', '&#039;').replace('"', '&quot;'),
                         cls: cls,
                         expanded: this.nodes[i].page_exists,
                         allowDrop: true,

```
